### PR TITLE
Remove side-effect free statements from interpreter trace.

### LINKED ISSUE #6786: [yul] interpreter should not leave trace of side-effects of code that will be removed by the optimizer
## Description

(Found by ossfuzz nightly CI)

```
{
let a_0, _1, a
function foo_0(x_0,_2, x_2) -> x_6
{
let x_4 := extcodehash(extcodesize(1))
}
a_0 := foo_0(calldataload(0),0, a_0)
}
```

is optimized to

```
{}
```

I wonder why this was not a problem earlier. Wouldn't the redundant assignment eliminator have found this earlier?

## Environment

- Compiler version: latest develop

### PR DESCRIPTION
**Pull request by chriseth**
Fixes https://github.com/ethereum/solidity/issues/6786

### CODE REVIEW

**Comment by bshastry** [comment]
`mload` is not excluded. This fails

```
{
  pop(f())
pop(f())
 pop(g())
pop(f())   pop(g())
  function f() -> bsl { let a := mload(0)}
  function g() -> a { let b := mload(5) a := b
//
  pop(f()) }
}
```

with

```
Unoptimized trace:
  MLOAD_FROM_SIZE(0, 32)
  MLOAD_FROM_SIZE(0, 32)
  MLOAD_FROM_SIZE(5, 32)
  MLOAD_FROM_SIZE(0, 32)
  MLOAD_FROM_SIZE(0, 32)
  MLOAD_FROM_SIZE(5, 32)
  MLOAD_FROM_SIZE(0, 32)

Optimized trace:
```

**Comment by bshastry** [comment]
Opcodes that should not have a side-effect but still do

- `mload`
- `keccak256`

File: test/libyul/yulInterpreterTests/side_effect_free.yul [hunk 1]
```diff
@@ -0,0 +1,16 @@
+{
+	// These can be removed by the optimizer and should not
+	// appear in the trace.
+	pop(gas())
+	pop(extcodesize(0))
+	pop(extcodehash(0))
+	pop(returndatasize())
+	pop(sload(0))
+	pop(pc())
+	pop(msize())
```
Comment (bshastry) [inline test/libyul/yulInterpreterTests/side_effect_free.yul:10]: Can be extended with

```
pop(mload(0))
pop(keccak256(0,0))
```

**Comment by bshastry** [comment]
some yul interpreter tests are failing

```
ASSERTION FAILURE:
- file   : boostTest.cpp
- line   : 125
- message: Fatal error during test.
Error parsing source.
:6:6: Error: Function not found.
	pop(extcodehash(0))
	    ^---------^
```

These functions cannot be found
  - `extcodehash()`
  - `returndatasize()`

**Comment by chriseth** [comment]
Combined with #6800

**Pull request closed without merging**